### PR TITLE
Add cu_seqlens support to OlmoHybridGatedDeltaNet for packed sequences

### PR DESCRIPTION
**Pull request by tyler-romero**
# What does this PR do?

Pass cu_seqlens derived from packed attention masks to FLA's ShortConvolution and chunk_gated_delta_rule kernels, preventing recurrent state from leaking across sequence boundaries during packed-sequence training.

Fixes # (issue)

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [x] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

File: src/transformers/models/olmo_hybrid/modeling_olmo_hybrid.py [hunk 1]
```diff
@@ -621,6 +621,24 @@ def torch_recurrent_gated_delta_rule(
 )
 
 
+def _cu_seqlens_from_packed_mask(attention_mask: torch.Tensor) -> torch.Tensor:
```
Comment (vasqu) [inline src/transformers/models/olmo_hybrid/modeling_olmo_hybrid.py:624]: I have to intervene a bit here. We should not manually unpad and pad - this is something that I explicitly disallowed in the model addition PR

We can still support padding-free training/forward tho by preparing the input properly before passing it the model. We are kinda stuck with FA supporting it with position ids only as well which was a mistake imo.

Tl;dr: Use our data collator https://github.com/huggingface/transformers/blob/16a5b0936dcaae6efbc03ab1a4fd98dc324bfb9e/src/transformers/data/data_collator.py#L1364
And pass the prepared input with its metadata (cu seq lens, max seq lens etc)

File: src/transformers/models/olmo_hybrid/modeling_olmo_hybrid.py [hunk 2]
```diff
@@ -719,14 +737,26 @@ def forward(
         attention_mask: torch.Tensor | None = None,
         **kwargs: Unpack[TransformersKwargs],
     ) -> torch.Tensor:
-        # Requires LEFT padding to work correctly
-        hidden_states = apply_mask_to_padding_states(hidden_states, attention_mask)
```
Comment (vasqu) [inline src/transformers/models/olmo_hybrid/modeling_olmo_hybrid.py:723]: This will stay then

File: src/transformers/models/olmo_hybrid/modeling_olmo_hybrid.py [hunk 3]
```diff
@@ -737,23 +767,23 @@ def forward(
...
-            q, cache=conv_state_q, use_precomputed=use_precomputed, output_final_state=use_cache
+            q, cache=conv_state_q, use_precomputed=use_precomputed, output_final_state=use_cache, cu_seqlens=cu_seqlens
         )
         k, new_conv_state_k = self.k_conv1d(
-            k, cache=conv_state_k, use_precomputed=use_precomputed, output_final_state=use_cache
+            k, cache=conv_state_k, use_precomputed=use_precomputed, output_final_state=use_cache, cu_seqlens=cu_seqlens
         )
         v, new_conv_state_v = self.v_conv1d(
-            v, cache=conv_state_v, use_precomputed=use_precomputed, output_final_state=use_cache
+            v, cache=conv_state_v, use_precomputed=use_precomputed, output_final_state=use_cache, cu_seqlens=cu_seqlens
```
Comment (vasqu) [inline src/transformers/models/olmo_hybrid/modeling_olmo_hybrid.py:776]: Should be hidden behind kwargs if possible (from the signature)

I hope FLA has **kwargs in its signature and it doesnt destroy anything

**Review by vasqu — COMMENTED** [review]
Commented on the modeling file (ofc would need to be modified in the modular file sry)

I just wanted to intervene because our approach is to do this before the model forward, not within. It always causes issues otherwise - be it compile, export, or just readability. I went into more detail with my first comment
